instrumentation/github.com/confluentinc/confluent-kafka-go/kafka/splunkkafka/consumer.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Splunk Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !windows
+// +build !windows
+
+package splunkkafka
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Consumer wraps a kafka.Consumer and traces its operations.
+type Consumer struct {
+	*kafka.Consumer
+	cfg    config
+	group  string
+	events chan kafka.Event
+	prev   trace.Span
+}
+
+// NewConsumer calls kafka.NewConsumer and wraps the resulting Consumer with
+// tracing instrumentation.
+func NewConsumer(conf *kafka.ConfigMap, opts ...Option) (*Consumer, error) {
+	c, err := kafka.NewConsumer(conf)
+	if err != nil {
+		return nil, err
+	}
+	// The kafka Consumer does not expose this. Give a best effort to add it.
+	var consumerGroup string
+	cGrpVal, err := conf.Get("group.id", "")
+	if err == nil {
+		consumerGroup, _ = cGrpVal.(string)
+	}
+	wrapped := &Consumer{
+		Consumer: c,
+		group:    consumerGroup,
+		cfg:      newConfig(opts...),
+	}
+	wrapped.events = wrapped.traceEventsChannel(c.Events())
+	return wrapped, nil
+}
+
+// WrapConsumer wraps a kafka.Consumer so that any consumed events are traced.
+func WrapConsumer(c *kafka.Consumer, opts ...Option) *Consumer {
+	wrapped := &Consumer{
+		Consumer: c,
+		cfg:      newConfig(opts...),
+	}
+	wrapped.events = wrapped.traceEventsChannel(c.Events())
+	return wrapped
+}
+
+func (c *Consumer) traceEventsChannel(in chan kafka.Event) chan kafka.Event {
+	// in will be nil when consuming via the events channel is not enabled.
+	if in == nil {
+		return nil
+	}
+
+	out := make(chan kafka.Event, cap(in))
+	go func() {
```

Comment:
should we not have here:

```go
		p.wg.Add(1)
		go func() {
			defer p.wg.Done()
```

and add `p.wg.Done()` in `Close` method?

Reply:
You can create an issue out of it if you do not want to spend time on this right now.